### PR TITLE
Hover tooltip for ImageStack glyph

### PR DESCRIPTION
**Pull request by ianthomas23**
Up until now the `"@image"` hover tooltip of an `ImageStack` glyph (#12356) has been the same as other image-like glyphs returning a single number corresponding to the data value of the array element under the cursor. This PR improves the tooltip by showing the slice of numbers from the 3D array that are under the cursor, which is a much better default.

Example:

small

produced by this code:
```python
from bokeh.models import ColorBar, EqHistColorMapper, WeightedStackColorMapper
from bokeh.palettes import varying_alpha_palette
from bokeh.plotting import figure, show
import numpy as np

colors = ["red", "blue", "green"]
stack = np.asarray([[[0, 1, 2], [1, 1, 2], [0, 1, 2]],
                    [[1, 0, 1], [2, 2, 2], [3, 3, 3]],
                    [[0, 1, 4], [7, 1, 3], [3, 5, 1]]])

p = figure(width=500, height=400, tooltips=[("values", "@image")])

alpha_mapper = EqHistColorMapper(palette=varying_alpha_palette(color="#000", n=10, start_alpha=40), rescale_discrete_levels=False)
color_mapper = WeightedStackColorMapper(palette=colors, nan_color=(0, 0, 0, 0), alpha_mapper=alpha_mapper)
p.image_stack(image=[stack], x=0, y=0, dw=1, dh=1, color_mapper=color_mapper)

color_bar = ColorBar(color_mapper=color_mapper)
p.add_layout(color_bar, "right")

show(p)
```
The source code change is generic and will apply to hovering over any form of 3D image, at the moment `ImageGlyph` is the only possibility of this.

The test code is derived from the very similar #11437 regression test code.

### CODE REVIEW

**Comment by jbednar** [comment]
Nice! The list representation is very compact (`values: 7,1,3`), but I wonder if a dictionary representation might be more usable/readable (`red: 7, green: 1, blue:3`).

**Comment by muendlein** [comment]
> Nice! The list representation is very compact (`values: 7,1,3`), but I wonder if a dictionary representation might be more usable/readable (`red: 7, green: 1, blue:3`).

Perhaps `r: 7, g: 1, b: 3` could provide the same information in a more compact form.

File: bokehjs/test/integration/glyphs/image_stack.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    function hover_at(plot_view: PlotView, r: Renderer, x: number, y: number) {
+      const crv = plot_view.renderer_views.get(r)!
+      const [[sx], [sy]] = crv.coordinates.map_to_screen([x], [y])
+
+      const ui = plot_view.canvas_view.ui_event_bus
+      const {left, top} = offset_bbox(ui.hit_area)
+
+      const ev = new MouseEvent("mousemove", {clientX: left + sx, clientY: top + sy})
+      ui.hit_area.dispatchEvent(ev)
+    }
```
Comment (mattpap) [inline bokehjs/test/integration/glyphs/image_stack.ts:44]: I would suggest looking into `PlotActions`, which should give you full control over canvas events without going into low level details like this. There plentiful examples of this in regression tests and a few other places.

Comment (ianthomas23) [reply bokehjs/test/integration/glyphs/image_stack.ts:44]: This is a copy-and-paste of the `describe("in issue #11437"` test which you authored. If it is good enough for that test then I assume it is good enough for this test.

**Comment by ianthomas23** [comment]
> Nice! The list representation is very compact (`values: 7,1,3`), but I wonder if a dictionary representation might be more usable/readable (`red: 7, green: 1, blue:3`).

It certainly won't be more readable, those colors are generally RGBA values e.g. `"#FA10B432"` and not suitable for use as dictionary keys presented to the user. The example happens to use `"red"`, `"green"` and `"blue"` but these are very unlikely to be used for a real use case.

It would be good to assign some sort of `name` or other metadata to each of the stacks in the `ImageStack`. That is very sensible but there isn't any functionality for that, and I am not attempting such in this PR.

File: bokehjs/test/integration/glyphs/image_stack.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/test/integration/glyphs/image_stack.ts:44]: That's an obsolete way of interactive testing and should have been replaced a long time ago. This applies to all tests that do this kind of low-level manipulation. That's why I'm pointing this out. There are much better ways of doing this now.

Comment (ianthomas23) [reply bokehjs/test/integration/glyphs/image_stack.ts:44]: Sure, I would rather use a better way.

**Comment by jbednar** [comment]
> The example happens to use "red", "green" and "blue" but these are very unlikely to be used for a real use case.
> It would be good to assign some sort of name or other metadata to each of the stacks in the ImageStack.

There seems to be some confusion here; I wasn't proposing that "red" or "green" or "blue" or RGBA hex codes have any special status; those just happen to be the names of the channels in this specific example, and my proposal is to use the name of each channel followed by the value in that channel. So, yes, the name would need to be preserved in the ImageStack; I don't know how we'll have sensible hover information without that.

**Comment by mattpap** [comment]
This information could be encoded within the `dtype`, though currently we don't support complex dtypes, especially dtypes with fields.

**Comment by ianthomas23** [comment]
> There seems to be some confusion here; I wasn't proposing that "red" or "green" or "blue" or RGBA hex codes have any special status; those just happen to be the names of the channels in this specific example, and my proposal is to use the name of each channel followed by the value in that channel. So, yes, the name would need to be preserved in the ImageStack; I don't know how we'll have sensible hover information without that.

I understand what you are asking for, but the channels do not have any names. There is just a 3D array of numbers and a palette of colors. So what you want is the addition of an extra property for the names, or something similar. I am not offering that in this PR, this is just a better default tooltip for 3D arrays of data. Subsequent PRs can add further improvements. Ideally I would like to see an issue created to specifically request what is desired.

**Review by bryevdv — APPROVED** [review]
This seems like a straightforward immediate improvement. I agree labeled arrays is a separate larger concern that would need to be addressed on its own.

**Pull request merged**